### PR TITLE
Adding api tests

### PR DESCRIPTION
**Pull request by yellowee**
Adding vitest package for API tests, updating CI and a test for auth token mutation

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migrations are either absent or optimized for zero downtime
* [ ] The changes are covered by test cases

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,53 +1,66 @@
 # Changelog
 
-All notable, unreleased changes to this project will be documented in this file. For the released changes, please visit the [Releases](https://github.com/mirumee/saleor/releases) page.
+All notable, unreleased changes to this project will be documented in this file. For the released changes, please visit
```
Comment (maarcingebala) [inline CHANGELOG.md:3]: Changelog contains unrelated formatting changes.

File: api-tests/package.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "@typescript-eslint/eslint-plugin": "5.59.6",
+    "@vitest/ui": "0.31.1",
+    "eslint-plugin-chai-friendly": "0.7.2",
+    "eslint-plugin-formatjs": "4.10.1",
+    "eslint-plugin-no-skip-tests": "1.1.0",
+    "eslint-plugin-prettier": "4.2.1",
+    "eslint-plugin-simple-import-sort": "10.0.0",
+    "prettier": "2.8.8",
+    "typescript": "5.0.4",
+    "vite": "4.3.8",
```
Comment (maarcingebala) [inline api-tests/package.json:33]: As @krzysztofzuraw originally suggested, I would also change pinned package versions to have automatic patches and minor updates instead. It helps to stay up-to-date with bug fixes and security releases.

Comment (yellowee) [reply api-tests/package.json]: https://github.com/saleor/saleor/pull/12963/commits/a3e557cbfa51372de84978926f7049e267d38b19

File: CHANGELOG.md (hunk 1 above)
Comment (yellowee) [reply CHANGELOG.md:3]: https://github.com/saleor/saleor/pull/12963/commits/cfabef3953c45a4d22cdb8c1eb8dd1b6960c00b8

File: api-tests/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    fragment AccountError on AccountError {
+      code
+      field
+      message
+    }
+  `
+  const result = await client.request<TokenCreateMutation>(mutation, {
+    email: 'testers+dashboard@saleor.io',
+    password: 'test1234',
```
Comment (maarcingebala) [inline api-tests/src/utils.ts:33]: Maybe credentials should also be moved to env variables?

Comment (yellowee) [reply api-tests/src/utils.ts]: https://github.com/saleor/saleor/pull/12963/commits/4c32afa1b18d91352509186d1d0c398f546069f9

File: api-tests/generated/fragment-masking.ts
Comment (fowczarek) [inline api-tests/generated/fragment-masking.ts:1]: I'm not sure that we should push generated files to the repository.

File: api-tests/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "api-tests",
+  "version": "1.0.0",
+  "private": true,
+  "description": "Saleor API tests",
+  "main": "N/A",
```
Comment (witoszekdev) [inline api-tests/package.json:6]: This will break if you `require()` this package - I think we should just create an `index.mjs` file and throw an error if someone tries to import it.

`N/A` is not a valid value

**Review by fowczarek — COMMENTED** [review]
There are a lot of duplicated commits. In the future consider using `git commit --amend` or interactive rebase to reach better commit history in branch.

**Review by fowczarek — COMMENTED** [review]
There are a lot of duplicated commits. In the future consider using `git commit --amend` or interactive rebase to reach better commit history in branch.

File: api-tests/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  "scripts": {
+    "test": "set -a && . ./.env && set +a && vitest",
```
Comment (witoszekdev) [inline api-tests/package.json:8]: We shouldn't expect our users to be running in a specific shell - `set -a` is UNIX only command

A typical practice is to use `dotenv` which handles loading `.env` files... or not loading `.env` files at all and using tools like [direnv](https://direnv.net/)

It looks like we're already using `dotenv` in vitest config: [api-tests/vitest.config.ts](https://github.com/saleor/saleor/pull/12963/files#diff-fdc5dd4df60190bc0e061ebfa9370a8e968c2da388af1fbbca79a73aa8dc11fc)

Moreover, `.env` file is not present in this PR, should we use `.env` as the default?

File: api-tests/.prettierrc
Comment (witoszekdev) [inline api-tests/.prettierrc:1]: We should avoid modifying default prettier config, here's a default one we use in other repos:

```
{
  "singleQuote": false,
  "printWidth": 100
}
```

File: api-tests/.env-example
Comment (witoszekdev) [inline api-tests/.env-example:1]: If this file is only an example, it should be named `.env` and the local overrides `.env.local`.

`.env-example` won't be recognized by most code editors as an `.env` file (it has a different syntax from pure bash/zsh script)

File: api-tests/package.json (hunk 4 above)
Comment (yellowee) [reply api-tests/package.json:6]: This change has been requested by @krzysztofzuraw in the original PR: https://github.com/saleor/saleor/pull/12944/files/cea5c85485d6fb2c62886672374a8da63fa0c59d#r1209163098 
Please decide among yourselves what approach should be used in this case and let me know once decided so I will make the changes as needed.

File: api-tests/package.json (hunk 5 above)
Comment (yellowee) [reply api-tests/package.json:8]: I've been using direnv in vitest.config.ts, but I changed my approach after @krzysztofzuraw 's suggestion in the original PR: https://github.com/saleor/saleor/pull/12944/files/cea5c85485d6fb2c62886672374a8da63fa0c59d#r1209154106

I've added .env file to .gitignore, but I provided /env-example, so everybody can set up env variables themselves - this way I don't need to hardcode credentials in my code. 

I've also added `  setupFiles: ['dotenv/config'],` to vitest.config.ts in order to use those variables.

If you still want me to make changes, please let me know.

File: api-tests/generated/fragment-masking.ts
Comment (yellowee) [reply api-tests/generated/fragment-masking.ts:1]: https://github.com/saleor/saleor/pull/12963/commits/0ad093e95de336f1170279668564b7f6ff1b7b13

File: api-tests/generated/gql.ts
[the remark this answers is not in the document]
Comment (yellowee) [reply api-tests/generated/gql.ts:1]: https://github.com/saleor/saleor/pull/12963/commits/0ad093e95de336f1170279668564b7f6ff1b7b13

Should I also add script for build as you suggested?

File: api-tests/.prettierrc
Comment (yellowee) [reply api-tests/.prettierrc:1]: https://github.com/saleor/saleor/pull/12963/commits/adf6c95bf7a7ceaf4f4b814bad51c20fa21503e1

File: api-tests/.gitignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+/node_modules
+.idea
```
[the remark this answers is not in the document]
Comment (yellowee) [reply api-tests/.gitignore:2]: https://github.com/saleor/saleor/commit/0ad093e95de336f1170279668564b7f6ff1b7b13

File: api-tests/.env-example
Comment (yellowee) [reply api-tests/.env-example:1]: https://github.com/saleor/saleor/pull/12963/commits/177439264375c44dfe9bf4088906659c3d53f412

File: api-tests/.prettierrc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "bracketSameLine": true,
+  "singleQuote": true,
+  "trailingComma": "all",
+  "semi": false,
```
[the remark this answers is not in the document]
Comment (yellowee) [reply api-tests/.prettierrc]: Modified to default values as suggested in https://github.com/saleor/saleor/pull/12963#discussion_r1210393984

https://github.com/saleor/saleor/commit/adf6c95bf7a7ceaf4f4b814bad51c20fa21503e1

File: api-tests/package.json (hunk 5 above)
Comment (krzysztofzuraw) [reply api-tests/package.json]: I think there are two types of env variables here:
1) used in tests - those can be covered by default vite env [handling](https://vitejs.dev/guide/env-and-mode.html)
2) used to generate GraphQL types based on schema - can't we use the same approach here as in [apps](https://github.com/saleor/saleor-app-template/blob/main/package.json#L10)?

File: CHANGELOG.md
[the remark this answers is not in the document]
Comment (yellowee) [reply CHANGELOG.md:1]: https://github.com/saleor/saleor/pull/12963/commits/a5d82ba2e53a3b8ffc5ec49682d1c3420324b0b4

File: api-tests/package.json (hunk 4 above)
Comment (yellowee) [reply api-tests/package.json]: I removed main from the package.json https://github.com/saleor/saleor/pull/12963/commits/55969f2944e6fa0c9a35c1e30cc13ac0b3034502

Let me know if any other changes are needed please, cheers!

File: .github/workflows/apiTests.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+      - run: npm ci
+        with:
+          node-version-file: '.nvmrc'
```
[the remark this answers is not in the document]
Comment (yellowee) [reply .github/workflows/apiTests.yml:15]: https://github.com/saleor/saleor/pull/12963/commits/c4366462eeeb8ce1a8cb6414935b06daba45840e

File: api-tests/package.json (hunk 5 above)
Comment (krzysztofzuraw) [reply api-tests/package.json]: Sure - idea there is that you have `schemaVersion` defined in pkg.json like this:
```json
  "saleor": {
    "schemaVersion": "3.10"
  },
```
Reference: https://github.com/saleor/saleor-app-template/blob/main/package.json#L14-L16

and then in `scripts` you can reference this variable as `npm_package_saleor_schemaVersion`

File: .github/workflows/apiTests.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+name: API-tests
+
+on:
+  pull_request:
```
Comment (NyanKiyoshi) [inline .github/workflows/apiTests.yml:4]: Did you mean `push`?

File: api-tests/package.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  "name": "api-tests",
+  "version": "1.0.0",
+  "private": true,
+  "description": "Saleor API tests",
+  "scripts": {
+    "test": "vitest",
+    "lint": "eslint 'src/**/*.@(tsx|ts|jsx|js)' --fix",
+    "check-types": "tsc --noEmit",
+    "generate": "graphql-codegen ",
+    "fetch-schema": "export $(cat .env.local| xargs) && curl $GRAPHQL_RAW > graphql/schema.graphql"
```
Comment (NyanKiyoshi) [inline api-tests/package.json:11]: That's a very dangerous way of getting `GRAPHQL_RAW`, you are already using a javascript eco-system here, simply write a safe `.js`  script to retrieve the schema

File: api-tests/graphql/schema.graphql
Comment (fowczarek) [inline api-tests/graphql/schema.graphql:1]: Do you need a second `schema.graphql` file in this repository?

File: api-tests/src/auth.test.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      mutation,
+      {
+        email,
+        password,
+      }
+    );
+
+    expect(result.tokenCreate?.csrfToken).toBeTypeOf("string");
+    expect(result.tokenCreate?.token).toBeTypeOf("string");
+    expect(result.tokenCreate?.refreshToken).toBeTypeOf("string");
```
Comment (fowczarek) [inline api-tests/src/auth.test.ts:43]: Shouldn't we check if those values working correctly? 
How its work when `token` will be an empty string(`""`)?

**Comment by yellowee** [comment]
Closing pull request due to a change of tech stack
